gapic-common/lib/gapic/rest/paged_enumerable.rb

```diff
@@ -0,0 +1,226 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Gapic
+  module Rest
+    ##
+    # A class to provide the Enumerable interface to the response of a REST paginated method.
+    # PagedEnumerable assumes response message holds a list of resources and the token to the next page.
+    #
+    # PagedEnumerable provides the enumerations over the resource data, and also provides the enumerations over the
+    # pages themselves.
+    #
+    # @example normal iteration over resources.
+    #   paged_enumerable.each { |resource| puts resource }
+    #
+    # @example per-page iteration.
+    #   paged_enumerable.each_page { |page| puts page }
+    #
+    # @example Enumerable over pages.
+    #   paged_enumerable.each_page do |page|
+    #     page.each { |resource| puts resource }
+    #   end
+    #
+    # @example more exact operations over pages.
+    #   while some_condition()
+    #     page = paged_enumerable.page
+    #     do_something(page)
+    #     break if paged_enumerable.next_page?
+    #     paged_enumerable.next_page
+    #   end
+    #
+    class PagedEnumerable
+      include Enumerable
+
+      ##
+      # @attribute [r] page
+      #   @return [Page] The current page object.
+      attr_reader :page
+
+      ##
+      # @private
+      # @param service_stub [Object] The REST service_stub with the baseline implementation for the wrapped method.
+      # @param method_name [Symbol] The REST method name that is being wrapped.
+      # @param request [Object] The request object.
+      # @param response [Object] The response object.
+      # @param options [Gapic::CallOptions] The options for making the RPC call.
+      # @param format_resource [Proc] A Proc object to format the resource object. The Proc should accept response as an
+      #   argument, and return a formatted resource object. Optional.
+      #
+      def initialize service_stub, method_name, resource_field_name, request, response, options, format_resource: nil
+        @service_stub = service_stub
+        @method_name = method_name
+        @resource_field_name = resource_field_name
+        @request = request
+        @response = response
+        @options = options
+        @format_resource = format_resource
+
+        @page = Page.new response, resource_field_name, format_resource: @format_resource
+      end
+
+      ##
+      # Iterate over the resources.
+      #
+      # @yield [Object] Gives the resource objects in the stream.
+      #
+      # @raise [RuntimeError] if it's not started yet.
```

Comment:
It doesn't look like this is actually implemented. Please either implement or remove.

Reply:
:+1: